### DISCUSSION TITLE
[BUG] PnP 2.* not working with Az-Module

### DISCUSSION THREAD

**Discussion by HeKr09** [pnp/powershell#3003] [Q&A] [answered]
### Reporting an Issue or Missing Feature
I am trying to run PnP and Azure in one PowerShell and receive an error.

### Expected behavior 
Both cmdlets should work

### Actual behavior 
I receive an error when I want to get an azure resource group after connecting with PnP.
![Error](https://user-images.githubusercontent.com/43664126/231197275-c9b7e143-9944-449e-94a3-94d4fc4184ac.PNG)

### Steps to reproduce behavior
```
Connect-PnPOnline -Url $WebUrl -Credentials $Credentials
Connect-AzAccount -Tenant "SOMEID" -Credential $Credentials | Out-Null
Connect-MgGraph -TenantId "SOMEID" -Scopes Group.ReadWrite.All | Out-Null

$AzureResourceGroupName = "SPW_RG"

# Removing Azure resource group
$resource = Get-AzResourceGroup -Name $AzureResourceGroupName

Disconnect-MgGraph
Disconnect-AzAccount
Disconnect-PnPOnline
```

### What is the version of the Cmdlet module you are running?
- Az: 9.4.0
- Az.Resources: 6.5.2
- PnP: 2.1.1 (also tested with 2.0.6-nightly) - but worked everything with 1.12.0

### Which operating system/environment are you running PnP PowerShell on?
- [x] Windows
- [ ] Linux
- [ ] MacOS
- [ ] Azure Cloud Shell
- [ ] Azure Functions
- [ ] Other : please specify

**Comment by veronicageek** [comment]
@HeKr09 - We did bump quite a lot of our underlying dependencies in the 2.1.1 update, so that might cause issues with other modules we have no control over. MSGraph and Az modules may be updating / upgrading their dependencies later on, but we would suggest running the PnP cmdlets in a different thread:

https://pnp.github.io/powershell/articles/azurefunctions.html

Moving to a discussion as it's unlikely we can do anything.

**Comment by AndersRask** [comment]
I did some digging on loaded assemblies for the mentioned module

`[System.AppDomain]::CurrentDomain.GetAssemblies() | ? location -like *az.res*`

It seems to be the assembly Microsoft.Extensions.Caching.Memory.dll (might be more than this) that is a much older version in the Az modules.

![image](https://user-images.githubusercontent.com/1651150/231747461-d1bc53b7-8d97-465d-b3a1-c4924d8e53e6.png)

**Reply by AndersRask** [reply]
@HeKr09 make sure you use Az.Resource first, by calling eg Get-AzResourceGroup *before* calling any PnP.PowerShell cmdlets, and it should work fine. The reason being that you will load the old assembly first, which works fine:

```
Connect-AzAccount -Tenant "SOMEID" -Credential $Credentials | Out-Null
Connect-MgGraph -TenantId "SOMEID" -Scopes Group.ReadWrite.All | Out-Null

$AzureResourceGroupName = "SPW_RG"

# Removing Azure resource group
$resource = Get-AzResourceGroup -Name $AzureResourceGroupName

# moved down
Connect-PnPOnline -Url $WebUrl -Credentials $Credentials
# use pnp here

Disconnect-MgGraph
Disconnect-AzAccount
Disconnect-PnPOnline
```

**Comment by HeKr09** [comment]
If I'll do this I receive the following error:
```
Connect-PnPOnline: D:\TFS\Remove-CustomerPortal.ps1:85
Line |
  85 |  Connect-PnPOnline -Url $WebUrl -Credentials $Credentials
     |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Could not load file or assembly 'System.IdentityModel.Tokens.Jwt, Version=6.27.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'. Could not find or load a specific file. (0x80131621)
Disconnect-PnPOnline: D:\TFS\Remove-SwoCustomerPortal.ps1:89
Line |
  89 |  Disconnect-PnPOnline
     |  ~~~~~~~~~~~~~~~~~~~~
     | No connection to disconnect
```

**Reply by AndersRask** [reply]
Ok, you need to do your graph stuff before also then :)

```
Connect-AzAccount -Tenant "SOMEID" -Credential $Credentials | Out-Null
Connect-MgGraph -TenantId "SOMEID" -Scopes Group.ReadWrite.All | Out-Null

$AzureResourceGroupName = "SPW_RG"

# Removing Azure resource group
$resource = Get-AzResourceGroup -Name $AzureResourceGroupName

# use graph for some stuff
Get-MGUser

# moved down
Connect-PnPOnline -Url $WebUrl -Credentials $Credentials
# use pnp here

Disconnect-MgGraph
Disconnect-AzAccount
Disconnect-PnPOnline
```

**Reply by HeKr09** [reply]
I do so. The posted script is not trimmed at all. I only do get the AzResource.
I don't do anything with Graph or PnP unless Connect and Disconnect.
I run the same order than yours.
1. Connect Az, Graph and PnP
2. Disconnect Graph, Az and PnP

**Reply by AndersRask** [reply]
hmm that is weird:
I had it working, but now it fails for me too with 
Connect-PnPOnline: Could not load file or assembly 'Microsoft.Identity.Client, Version=4.50.0.0, Culture=neutral, PublicKeyToken=0a613f4dd989e8ae'. Could not find or load a specific file. (0x80131621)

It seems there is no way of enforcing load of newer versions of assemblies in a runspace unfortunately!
Even when I get it loaded using
`[System.Reflection.Assembly]::LoadFile("C:\yourpath\PowerShell\Modules\PnP.PowerShell\2.1.1\Core\Microsoft.Identity.Client.dll")`

Connect-PnPOnline still fails with same error, even though I can see the assembly is loaded:
`[System.AppDomain]::CurrentDomain.GetAssemblies() | ? location -like *identity.client* | fl fullname`

**Reply by HeKr09** [reply]
@AndersRask: Is there a chance that you submit an issue at the community? I guess you know much more with these dependencies than I do.

**Reply by AndersRask** [reply, marked as answer]
as @veronicageek mentions, this is a known issue that is hard to avoid building with new versions of assemblies, when other modules do similar things (Az and graph).
It is for sure annoying, and will affect adoption of the new module, but not much to do about it i'm afraid.
